# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: third_party/rust/cubeb-coreaudio/README.md
# Commit: 24b97b0c24f0
# Full Hash: 24b97b0c24f0f30965d5d30dfa2a796e3c994e5e
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2020-02-13 21:42:57
# Description:
#   Bug 1613156 - Update cubeb-coreaudio to 7fe03b4. r=padenot
#   
#   Pick commits:
#   - 7fe03b4: Bail out if the output device has no output channel (#50)
#   - ad74bad: Update README
# ==============================================================================

diff -r 8e18b9d4caae -r 24b97b0c24f0 third_party/rust/cubeb-coreaudio/README.md
--- a/third_party/rust/cubeb-coreaudio/README.md	Thu Feb 13 16:09:09 2020 +0000
+++ b/third_party/rust/cubeb-coreaudio/README.md	Thu Feb 13 15:52:30 2020 +0000
@@ -5,12 +5,13 @@
 *Rust* implementation of [Cubeb][cubeb] on [the MacOS platform][cubeb-au].
 
 ## Current Goals
+
 - Keep refactoring the implementation until it looks rusty! (it's translated from C at first.)
   - Check the [todo list][todo] first
 
 ## Status
 
-The code is currently tested in the _Firefox Nightly_ under a _perf_.
+The code is currently shipped within the _Firefox Nightly_ under a _perf_.
 
 - Try it:
   - Open `about:config`
@@ -21,6 +22,7 @@
   - Retart Firefox Nightly again if it's not.
 
 ## Test
+
 Please run `sh run_tests.sh`.
 
 Some tests cannot be run in parallel.
@@ -33,18 +35,29 @@
 Most of the tests are executed in `run_tests.sh`.
 Only those tests commented with *FIXIT* are left.
 
+### Run Sanitizers
+
+Run _AddressSanitizer (ASan), LeakSanitizer (LSan), MemorySanitizer (MSan), ThreadSanitizer (TSan)_
+by `sh run_sanitizers.sh`.
+
+The above command will run all the test suits in *run_tests.sh* by all the available _sanitizers_.
+However, it takes a long time for finshing the tests.
+
 ### Device Switching
+
 The system default device will be changed during our tests.
 All the available devices will take turns being the system default device.
 However, after finishing the tests, the default device will be set to the original one.
 The sounds in the tests should be able to continue whatever the system default device is.
 
 ### Device Plugging/Unplugging
+
 We implement APIs simulating plugging or unplugging a device
 by adding or removing an aggregate device programmatically.
 It's used to verify our callbacks for minitoring the system devices work.
 
 ### Manual Test
+
 - Output devices switching
   - `$ cargo test test_switch_output_device -- --ignored --nocapture`
   - Enter `s` to switch output devices
@@ -57,24 +70,18 @@
   - Plug/Unplug devices to see events log.
 
 ## TODO
+
 See [todo list][todo]
 
 ## Issues
+
 - Atomic:
   - We need atomic type around `f32` but there is no this type in the stardard Rust
   - Using [atomic-rs](https://github.com/Amanieu/atomic-rs) to do this.
-- No guarantee on `audiounit_set_channel_layout`
-  - This call doesn't work all the times
-  - Returned `NO_ERR` doesn't guarantee the layout is set to the one we want
-  - The layouts on some devices won't be changed even no errors are returned,
-    e.g., we can set _stereo_ layout to a _4-channels aggregate device_ with _QUAD_ layout
-    (created by Audio MIDI Setup) without any error. However, the layout
-    of this 4-channels aggregate device is still QUAD after setting it without error
-  - Another weird thing is that we will get a `kAudioUnitErr_InvalidPropertyValue`
-    if we set the layout to _QUAD_. It's the same layout as its original one but it cannot be set!
 - `kAudioDevicePropertyBufferFrameSize` cannot be set when another stream using the same device with smaller buffer size is active. See [here][chg-buf-sz] for details.
 
 ### Test issues
+
 - Fail to run tests that depend on `AggregateDevice::create_blank_device` with the tests that work with the device event listeners
   - The `AggregateDevice::create_blank_device` will add an aggregate device to the system and fire the device-change events indirectly.
 - `TestDeviceSwitcher` cannot work when there is an alive full-duplex stream
@@ -86,6 +93,7 @@
   - See details in [device_change.rs](src/backend/tests/device_change.rs)
 
 ## Branches
+
 - [trailblazer][trailblazer]: Main branch
 - [plain-translation-from-c][from-c]: The code is rewritten from C code on a line-by-line basis
 - [ocs-disposal][ocs-disposal]: The first version that replace our custom mutex by Rust Mutex